### ISSUE TITLE
Exceptions from HaskellExternalAnnotator

### ISSUE THREAD

**Issue by hoha** [carymrobbins/intellij-haskforce#288]
Several reports like this one have been reported IntelliJ exception analyzer tool. 

```
Read access is allowed from event dispatch thread or inside read-action only (see com.intellij.openapi.application.Application.runReadAction())
Details: Current thread: Thread[ApplicationImpl pooled thread 7,4,main] 489265438
; dispatch thread: false; isDispatchThread(): false
SystemEventQueueThread: Thread[AWT-EventQueue-0 2016.2.1#IU-162.1447.7 IDEA, eap:true, os:Mac OS X 10.11.5, java-version:JetBrains s.r.o 1.8.0_76-release-b216,6,main] 1494074506

java.lang.Throwable
    at com.intellij.openapi.diagnostic.Logger.error(Logger.java:138)
    at com.intellij.openapi.application.impl.ApplicationImpl.assertReadAccessAllowed(ApplicationImpl.java:950)
    at com.intellij.psi.impl.file.impl.FileManagerImpl.findFile(FileManagerImpl.java:385)
    at com.intellij.psi.impl.PsiManagerImpl.findFile(PsiManagerImpl.java:185)
    at com.haskforce.cabal.completion.CabalFileFinder$.com$haskforce$cabal$completion$CabalFileFinder$$getPsi(CabalFileFinder.scala:36)
    at com.haskforce.cabal.completion.CabalFileFinder$$anonfun$psiForFile$1.apply(CabalFileFinder.scala:32)
    at com.haskforce.cabal.completion.CabalFileFinder$$anonfun$psiForFile$1.apply(CabalFileFinder.scala:32)
    at scala.Option.flatMap(Option.scala:171)
    at com.haskforce.cabal.completion.CabalFileFinder$.psiForFile(CabalFileFinder.scala:32)
    at com.haskforce.cabal.query.BuildInfoUtil$.getBuildInfo(BuildInfoUtil.scala:12)
    at com.haskforce.cabal.query.BuildInfoUtil.getBuildInfo(BuildInfoUtil.scala)
    at com.haskforce.highlighting.annotation.external.HLint.getParamsFromCabal(HLint.java:118)
    at com.haskforce.highlighting.annotation.external.HLint.getParams(HLint.java:113)
    at com.haskforce.highlighting.annotation.external.HLint.access$100(HLint.java:45)
    at com.haskforce.highlighting.annotation.external.HLint$3.apply(HLint.java:91)
    at com.haskforce.highlighting.annotation.external.HLint$3.apply(HLint.java:87)
    at scala.util.Either.fold(Either.scala:99)
    at com.haskforce.highlighting.annotation.external.HLint.parseProblems(HLint.java:80)
    at com.haskforce.highlighting.annotation.external.HLint.lint(HLint.java:58)
    at com.haskforce.highlighting.annotation.external.impl.HLintProblemsProvider.getProblems(HLintProblemsProvider.scala:24)
    at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1$$anonfun$apply$1.apply(HaskellExternalAnnotator.scala:94)
    at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1$$anonfun$apply$1.apply(HaskellExternalAnnotator.scala:94)
    at scala.Option.map(Option.scala:146)
    at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1.apply(HaskellExternalAnnotator.scala:94)
    at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1.apply(HaskellExternalAnnotator.scala:94)
    at scala.collection.immutable.List.flatMap(List.scala:327)
    at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$.buildForFile(HaskellExternalAnnotator.scala:94)
    at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.doAnnotate(HaskellExternalAnnotator.scala:44)
    at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.doAnnotate(HaskellExternalAnnotator.scala:19)
    at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:212)
    at com.intellij.codeInsight.daemon.impl.ExternalToolPass.access$500(ExternalToolPass.java:44)
    at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.run(ExternalToolPass.java:160)
    at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:320)
    at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:310)
    at com.intellij.util.ui.update.MergingUpdateQueue.lambda$flush$1(MergingUpdateQueue.java:260)
    at com.intellij.util.ui.update.MergingUpdateQueue.flush(MergingUpdateQueue.java:274)
    at com.intellij.util.ui.update.MergingUpdateQueue.run(MergingUpdateQueue.java:229)
    at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:222)
    at com.intellij.util.Alarm$Request$1.run(Alarm.java:378)
    at com.intellij.util.Alarm$Request.run(Alarm.java:389)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:227)
    at com.intellij.util.concurrency.BoundedTaskExecutor.runFirstTaskThenPollAndRunRest(BoundedTaskExecutor.java:178)
    at com.intellij.util.concurrency.BoundedTaskExecutor.access$000(BoundedTaskExecutor.java:40)
    at com.intellij.util.concurrency.BoundedTaskExecutor$2.run(BoundedTaskExecutor.java:197)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by carymrobbins** [comment]
I bet this is covered by the fix introduced for #271. I'll work on getting a release out soon so you can confirm the fix.

**Comment by carymrobbins** [comment]
The release is out, so please report back if you encounter this issue in the future. Thanks!

**Comment by pnaranja** [comment]
Just updated to 0.3-beta.36

Also using:
Windows 10
Java 1.8.0_92
IntelliJ IDEA 2016.2 Community Edition


I see a similar exception:

```
Read access is allowed from event dispatch thread or inside read-action only (see com.intellij.openapi.application.Application.runReadAction())
Details: Current thread: Thread[ApplicationImpl pooled thread 19,4,main] 21937974
; dispatch thread: false; isDispatchThread(): false
SystemEventQueueThread: Thread[AWT-EventQueue-0 2016.2.1#IC-162.1447.26 IDEA, eap:false, os:Windows 10 10.0, java-version:Oracle Corporation 1.8.0_92-b14,6,main] 660705910
java.lang.Throwable
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:138)
	at com.intellij.openapi.application.impl.ApplicationImpl.assertReadAccessAllowed(ApplicationImpl.java:951)
	at com.intellij.psi.impl.source.tree.CompositeElement.getChildrenAsPsiElements(CompositeElement.java:475)
	at com.intellij.psi.impl.source.PsiFileImpl.getChildren(PsiFileImpl.java:830)
	at com.haskforce.cabal.query.CabalQuery.getLibrary(CabalQuery.scala:38)
	at com.haskforce.cabal.query.CabalQuery.getBuildInfo(CabalQuery.scala:56)
	at com.haskforce.cabal.query.CabalQuery$$anonfun$findBuildInfoForSourceFile$1.apply(CabalQuery.scala:65)
	at com.haskforce.cabal.query.CabalQuery$$anonfun$findBuildInfoForSourceFile$1.apply(CabalQuery.scala:64)
	at scala.Option.flatMap(Option.scala:171)
	at com.haskforce.cabal.query.CabalQuery.findBuildInfoForSourceFile(CabalQuery.scala:64)
	at com.haskforce.cabal.query.CabalQuery$$anonfun$findBuildInfoForSourceFile$3.apply(CabalQuery.scala:74)
	at com.haskforce.cabal.query.CabalQuery$$anonfun$findBuildInfoForSourceFile$3.apply(CabalQuery.scala:74)
	at scala.Option.flatMap(Option.scala:171)
	at com.haskforce.cabal.query.CabalQuery.findBuildInfoForSourceFile(CabalQuery.scala:74)
	at com.haskforce.cabal.query.CabalQuery$$anonfun$findBuildInfoForSourceFile$2.apply(CabalQuery.scala:70)
	at com.haskforce.cabal.query.CabalQuery$$anonfun$findBuildInfoForSourceFile$2.apply(CabalQuery.scala:70)
	at scala.Option.flatMap(Option.scala:171)
	at com.haskforce.cabal.query.CabalQuery.findBuildInfoForSourceFile(CabalQuery.scala:70)
	at com.haskforce.cabal.query.BuildInfoUtil$$anonfun$getBuildInfo$2.apply(BuildInfoUtil.scala:14)
	at com.haskforce.cabal.query.BuildInfoUtil$$anonfun$getBuildInfo$2.apply(BuildInfoUtil.scala:12)
	at scala.Option.flatMap(Option.scala:171)
	at com.haskforce.cabal.query.BuildInfoUtil$.getBuildInfo(BuildInfoUtil.scala:12)
	at com.haskforce.cabal.query.BuildInfoUtil.getBuildInfo(BuildInfoUtil.scala)
	at com.haskforce.highlighting.annotation.external.HLint.getParamsFromCabal(HLint.java:118)
	at com.haskforce.highlighting.annotation.external.HLint.getParams(HLint.java:113)
	at com.haskforce.highlighting.annotation.external.HLint.access$100(HLint.java:45)
	at com.haskforce.highlighting.annotation.external.HLint$3.apply(HLint.java:91)
	at com.haskforce.highlighting.annotation.external.HLint$3.apply(HLint.java:87)
	at scala.util.Either.fold(Either.scala:99)
	at com.haskforce.highlighting.annotation.external.HLint.parseProblems(HLint.java:80)
	at com.haskforce.highlighting.annotation.external.HLint.lint(HLint.java:58)
	at com.haskforce.highlighting.annotation.external.impl.HLintProblemsProvider.getProblems(HLintProblemsProvider.scala:24)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1$$anonfun$apply$1.apply(HaskellExternalAnnotator.scala:94)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1$$anonfun$apply$1.apply(HaskellExternalAnnotator.scala:94)
	at scala.Option.map(Option.scala:146)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1.apply(HaskellExternalAnnotator.scala:94)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1.apply(HaskellExternalAnnotator.scala:94)
	at scala.collection.immutable.List.flatMap(List.scala:327)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$.buildForFile(HaskellExternalAnnotator.scala:94)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.doAnnotate(HaskellExternalAnnotator.scala:44)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.doAnnotate(HaskellExternalAnnotator.scala:19)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:212)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.access$500(ExternalToolPass.java:44)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.run(ExternalToolPass.java:160)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:320)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:310)
	at com.intellij.util.ui.update.MergingUpdateQueue.lambda$flush$1(MergingUpdateQueue.java:260)
	at com.intellij.util.ui.update.MergingUpdateQueue.flush(MergingUpdateQueue.java:274)
	at com.intellij.util.ui.update.MergingUpdateQueue.run(MergingUpdateQueue.java:229)
	at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:222)
	at com.intellij.util.Alarm$Request$1.run(Alarm.java:378)
	at com.intellij.util.Alarm$Request.run(Alarm.java:389)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:227)
	at com.intellij.util.concurrency.BoundedTaskExecutor.runFirstTaskThenPollAndRunRest(BoundedTaskExecutor.java:178)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$000(BoundedTaskExecutor.java:40)
	at com.intellij.util.concurrency.BoundedTaskExecutor$2.run(BoundedTaskExecutor.java:197)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by carymrobbins** [comment]
Thanks, I'll get this worked out.

**Comment by davips** [comment]
Same in Debian Jessie.

**Comment by carymrobbins** [comment]
Another one

```
2017-08-02 19:32:10,405 [39693633]  ERROR - plication.TransactionGuardImpl - Write-unsafe context! Model changes are allowed from write-safe contexts only. Please ensure you're using invokeLater/invokeAndWait with a correct modality state (not "any"). See TransactionGuard documentation for details.
  current modality=ModalityState.NON_MODAL
  known modalities={ModalityState:Loading Project 494019902: running=false; canceled=false=true, ModalityState.NON_MODAL=true, ModalityState:com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog[dialog7,331,177,1037x735,invalid,hidden,layout=java.awt.BorderLayout,APPLICATION_MODAL,title=Project Structure,defaultCloseOperation=DO_NOTHING_ON_CLOSE,rootPane=,rootPaneCheckingEnabled=true]=true, ModalityState:=true, ModalityState:com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog[dialog10,331,177,1037x735,invalid,hidden,layout=java.awt.BorderLayout,APPLICATION_MODAL,title=Project Structure,defaultCloseOperation=DO_NOTHING_ON_CLOSE,rootPane=,rootPaneCheckingEnabled=true]=true, ModalityState:com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog[dialog4,331,177,1037x735,invalid,hidden,layout=java.awt.BorderLayout,APPLICATION_MODAL,title=Project Structure,defaultCloseOperation=DO_NOTHING_ON_CLOSE,rootPane=,rootPaneCheckingEnabled=true]=true, ModalityState:com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog[dialog7,331,177,1037x735,invalid,hidden,layout=java.awt.BorderLayout,APPLICATION_MODAL,title=Project Structure,defaultCloseOperation=DO_NOTHING_ON_CLOSE,rootPane=,rootPaneCheckingEnabled=true], com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog[dialog8,592,36,1080x635,invalid,hidden,layout=java.awt.BorderLayout,APPLICATION_MODAL,title=New Module,defaultCloseOperation=DO_NOTHING_ON_CLOSE,rootPane=,rootPaneCheckingEnabled=true]=true, ModalityState:com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog[dialog6,331,177,1037x735,invalid,hidden,layout=java.awt.BorderLayout,APPLICATION_MODAL,title=Project Structure,defaultCloseOperation=DO_NOTHING_ON_CLOSE,rootPane=,rootPaneCheckingEnabled=true]=true}
java.lang.Throwable
        at com.intellij.openapi.diagnostic.Logger.error(Logger.java:132)
        at com.intellij.openapi.application.TransactionGuardImpl.assertWriteActionAllowed(TransactionGuardImpl.java:230)
        at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.saveAllDocuments(FileDocumentManagerImpl.java:298)
        at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.saveAllDocuments(FileDocumentManagerImpl.java:290)
        at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$$anonfun$saveAllFiles$1.apply$mcV$sp(HaskellExternalAnnotator.scala:58)
        at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$$anonfun$saveAllFiles$1.apply(HaskellExternalAnnotator.scala:58)
        at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$$anonfun$saveAllFiles$1.apply(HaskellExternalAnnotator.scala:58)
        at com.haskforce.Implicits$Fun0$$anon$3.run(Implicits.scala:16)
        at com.intellij.openapi.application.impl.LaterInvocator$1.run(LaterInvocator.java:163)
        at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.a(LaterInvocator.java:416)
        at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:399)
        at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
        at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:762)
        at java.awt.EventQueue.access$500(EventQueue.java:98)
        at java.awt.EventQueue$3.run(EventQueue.java:715)
        at java.awt.EventQueue$3.run(EventQueue.java:709)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
        at java.awt.EventQueue.dispatchEvent(EventQueue.java:732)
        at com.intellij.ide.IdeEventQueue.e(IdeEventQueue.java:821)
        at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:649)
        at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:365)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
        at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
        at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by carymrobbins** [comment]
See this issue for another stack trace which appears to be related - https://github.com/carymrobbins/intellij-haskforce/issues/327

**Comment by carymrobbins** [comment]
This should be resolved in the next release, `0.3.39`
